algo/src/main/java/org/neo4j/graphalgo/similarity/SparseWeightedInput.java

```diff
@@ -0,0 +1,98 @@
+package org.neo4j.graphalgo.similarity;
+
+import org.neo4j.graphalgo.core.utils.Intersections;
+
+class SparseWeightedInput implements WeightedInput {
+    private final int initialSize;
+    private long id;
+    double[] weights;
+    private int count;
+    private RleDecoder rleDecoder;
+
+    public SparseWeightedInput(long id, double[] weights, int initialSize, int nonSkipSize) {
+        this.initialSize = initialSize;
+        this.id = id;
+        this.weights = weights;
+        this.count = nonSkipSize;
+        rleDecoder = new RleDecoder(initialSize);
+    }
+
+    public int compareTo(WeightedInput o) {
+        return Long.compare(id(), o.id());
+    }
+
+    public SimilarityResult sumSquareDeltaSkip(double similarityCutoff, WeightedInput other, double skipValue) {
+        int len = Math.min(weights().length, other.weights().length);
+        double sumSquareDelta = Intersections.sumSquareDeltaSkip(weights(), other.weights(), len, skipValue);
+        long intersection = 0;
+        /* todo
+        for (int i = 0; i < len; i++) {
+            if (weights()[i] == other.weights()[i] && weights()[i] != 0d) intersection++;
+        }
+        */
+        if (similarityCutoff >= 0d && sumSquareDelta > similarityCutoff) return null;
+        return new SimilarityResult(id(), other.id(), count(), other.count(), intersection, sumSquareDelta);
+    }
+
+    public SimilarityResult sumSquareDelta(double similarityCutoff, WeightedInput other) {
+        int len = Math.min(weights().length, other.weights().length);
+        double sumSquareDelta = Intersections.sumSquareDelta(weights(), other.weights(), len);
+        long intersection = 0;
+        /* todo
+        for (int i = 0; i < len; i++) {
+            if (weights()[i] == other.weights()[i] && weights()[i] != 0d) intersection++;
+        }
+        */
+        if (similarityCutoff >= 0d && sumSquareDelta > similarityCutoff) return null;
+        return new SimilarityResult(id(), other.id(), count(), other.count(), intersection, sumSquareDelta);
+    }
+
+    public SimilarityResult cosineSquaresSkip(double similarityCutoff, WeightedInput other, double skipValue) {
+        rleDecoder.reset(this.weights(), other.weights());
+
+        double[] thisWeights = rleDecoder.item1();
```

Comment:
(shouldn't we do this on the fly without decoding the whole array?)
to save decoding the same array twice pass the decoder from the outside.

Reply:
one RLEDecoder per work thread basically.

Reply:
we can also pass in a "Decoder" (or null, or an null-decoder) for the methods.